### PR TITLE
unix,windows: init all requests in fs calls

### PR DESCRIPTION
**Pull request by cjihrig**
Prior to this change, several of the fs functions checked for invalid arguments before initializing the fs request. If a consumer received a UV_EINVAL from one of these functions, and then called uv_fs_req_cleanup(), the application would crash, as the pointer being freed was not allocated. This commit makes sure that all fs functions initialize the request before returning.

Fixes: https://github.com/libuv/libuv/issues/1508

### CODE REVIEW

File: test/test-fs.c [hunk 1]
```diff
@@ -2751,19 +2751,23 @@ TEST_IMPL(fs_write_alotof_bufs_with_offset) {
 TEST_IMPL(fs_read_write_null_arguments) {
   int r;
 
-  r = uv_fs_read(NULL, NULL, 0, NULL, 0, -1, NULL);
+  r = uv_fs_read(NULL, &read_req, 0, NULL, 0, -1, NULL);
```
Comment (saghul) [inline test/test-fs.c]: hum, what should happen now if someone passes a NULL pointer then?

Comment (saghul) [reply test/test-fs.c]: I guess nothing? Since uv_fs_cleanup would also do nothing if a NULL pointer is passed?

Comment (cjihrig) [reply test/test-fs.c]: It crashes because none of the req init logic checks for `NULL`. That would happen for other calls too though. I think we should either crash consistently or update the init logic to handle `NULL`.

**Comment by cjihrig** [comment]
FWIW, green CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/410/

File: test/test-fs.c (hunk 1 above)
Comment (cjihrig) [reply test/test-fs.c]: So the more I think about this, the more I think we should just assert on req for these functions. If we return `UV_EINVAL` in these cases, there is no way for the consumer to know if they can cleanup the request or not. It also seems pretty nonsensical to call any of these without a request.

Comment (cjihrig) [reply test/test-fs.c]: @saghul I added a second commit to support `NULL` requests for all of the fs functions. The new logic is "if the request is `NULL`, immediately return `UV_EINVAL`. Otherwise, initialize the request before doing anything else." I think that should allow all requests to be cleaned up. PTAL

**Comment by cjihrig** [comment]
@saghul ping

File: src/unix/fs.c [hunk 2]
```diff
@@ -66,6 +66,8 @@
 
 #define INIT(subtype)                                                         \
   do {                                                                        \
+    if (req == NULL)                                                          \
```
Comment (saghul) [inline src/unix/fs.c]: Then `uv_fs_req_cleanup` should check if req is null and not do anything: https://github.com/libuv/libuv/blob/v1.x/src/unix/fs.c#L1452

Comment (saghul) [reply src/unix/fs.c]: Same on Windows.

File: src/win/fs.c [hunk 3]
```diff
@@ -1989,6 +1992,9 @@ int uv_fs_open(uv_loop_t* loop, uv_fs_t* req, const char* path, int flags,
 
 
 int uv_fs_close(uv_loop_t* loop, uv_fs_t* req, uv_file fd, uv_fs_cb cb) {
+  if (req == NULL)
```
Comment (saghul) [inline src/win/fs.c]: I see every function looks like this now on Windows:

```
if (req == NULL)
  return UV_EINVAL;

 uv_fs_req_init(loop, req, UV_FS_FOO, cb);
```

Maybe use a macro to save some lines?

File: test/test-fs.c (hunk 1 above)
Comment (saghul) [reply test/test-fs.c]: Please also add some tests which use a NULL req.

**Comment by cjihrig** [comment]
@saghul addressed all of your comments.

CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/422/

**Comment by cjihrig** [comment]
Squashed and landed as 7a0e64d2e0e6a01673365a4062067a79ff690712, e539fc412f87ec215bd7291c3fdf36aed7557667, 9a4468f47aead2c9fad1735ebcdcf51d160c860e. Thanks!
